osu.Game/Lists/IntervalList.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using osu.Framework.Lists;
+
+namespace osu.Game.Lists
+{
+    /// <summary>
+    /// Represents a list of intervals that can be used for whether a specific value falls into one of them.
+    /// </summary>
+    /// <typeparam name="T">The type of interval values.</typeparam>
+    public class IntervalList<T> : IEnumerable<Interval<T>>
+        where T : struct, IConvertible
+    {
+        private static readonly IComparer<T> type_comparer = Comparer<T>.Default;
+
+        private readonly SortedList<Interval<T>> intervals = new SortedList<Interval<T>>((x, y) => type_comparer.Compare(x.Start, y.Start));
+        private int nearestIndex;
+
+        public Interval<T> this[int i]
+        {
+            get => intervals[i];
+            set => intervals[i] = value;
+        }
```

Comment:
Not required and potentially buggy (consider what happens if you do
```csharp
intervalList[1] = new Interval<double>(15, 20);
```
after construction - will it break? will it be fine? I don't know, but it shouldn't be possible in the first place)